### ISSUE TITLE
building iOS no longer working after updating XCode from 8.2 to 8.3

### ISSUE THREAD

**Issue by strille** [Microsoft/remotebuild#26]
When running remotebuild (version 2.1.3) on a Mac with XCode 8.2, a deprecation warning was shown:

> PackageApplication is deprecated, use `xcodebuild -exportArchive` instead.

After updating to XCode 8.3 it seems this feature was removed altogether, because now the log shows:

> xcrun: error: unable to find utility "PackageApplication", not a developer tool or in PATH

It looks like remotebuild needs to be updated to use xcodebuild instead of xcrun.

**Comment by domisgigner** [comment]
I agree changing _src > taco-remote-lib > ios > iosBuild.ts_ line 90 from

`child_process.exec("xcodebuild -sdk iphoneos -exportArchive -archivePath" + pathToCordovaApp + " -exportPath " + fullPathToIpaFile, {},`

to

`child_process.exec("xcodebuild -sdk iphoneos -exportArchive -archivePath" + pathToCordovaApp + " -exportPath " + fullPathToIpaFile, {},
`
then recompiling and installing back on the mac worked for me (although the plist is now always generated as exportOptions.plist instead of appName.plist

**Comment by fahms** [comment]
Just ran into the same issue. I am using TFS build step to create package and ran into this issue:
`xcrun: error: unable to find utility "PackageApplication", not a developer tool or in PATH
/usr/bin/xcrun failed with return code: 72
Return code: 1 `

**Comment by Willovent** [comment]
This is really annoying since with Xcode 8.2.x we can't deploy on IOS 10.3.x and with Xcode 8.3 remotebuild doesn't work.

**Comment by drewgillies** [comment]
This is our number one issue right now, @Willovent. No ETA yet, but we're on it. Totally understand the annoyance factor.

**Comment by AnBucyk** [comment]
Fixed in PR #27, But there's still some process to go through to get it localized and published.

**Comment by darrenschwarz** [comment]
I am relatively new to IOS development, and having spent the time setting up my provisioning profile, configuring remotebuild, fixing the issues with xcode recognising my provisioning profile, I thought I was there when VS output showed 1>  ** BUILD SUCCEEDED **., only to get....

xcrun : error : unable to find utility "PackageApplication", not a developer tool or in PATH

I am using Visual Studio, to deploy a Cordova project, via remotebuild , on OS 10.3.x with Xcode 8.3.

Can you give a brief explanation as to what will need updating / how to go about it, when a fix is released. Thanks

**Comment by Willovent** [comment]
Well, you juste have to run this command to update remotebuild : `npm install --global remotebuild` on the MAC machine

**Comment by DarrylLeckenby** [comment]
Attempting to address this issue I have ran a remotebuild update on the remote build mac
![image](https://cloud.githubusercontent.com/assets/3273924/25628563/9308a412-2f5e-11e7-8713-79b5e93d223d.png)

However I am still seeing this issue
```
1>  ** BUILD SUCCEEDED **
1>
1>xcrun : error : unable to find utility "PackageApplication", not a developer tool or in PATH
1>  Failed to build app for buildNumber 741: Error code 72 for command: xcrun with args: -sdk,iphoneos,PackageApplication
1>  ,-v,/Users/developer/.taco_home/remote-builds/taco-remote/builds/741/cordovaApp/platforms/ios/build/device/Huntfun.ap
1>  p,-o,/Users/developer/.taco_home/remote-builds/taco-remote/builds/741/cordovaApp/platforms/ios/build/device/Huntfun.i
1>  pa
1>MSBUILD : cordova-build error : Error: Remote build error from the build server Build failed with error Error code 72 f
1>or command: xcrun with args: -sdk,iphoneos,PackageApplication,-v,/Users/developer/.taco_home/remote-builds/taco-remote/
1>builds/741/cordovaApp/platforms/ios/build/device/Huntfun.app,-o,/Users/developer/.taco_home/remote-builds/taco-remote/b
1>uilds/741/cordovaApp/platforms/ios/build/device/Huntfun.ipa - 1
1>  Error Remote build error from the build server Build failed with error Error code 72 for command: xcrun with args: -s
1>  dk,iphoneos,PackageApplication,-v,/Users/developer/.taco_home/remote-builds/taco-remote/builds/741/cordovaApp/platfor
1>  ms/ios/build/device/Huntfun.app,-o,/Users/developer/.taco_home/remote-builds/taco-remote/builds/741/cordovaApp/platfo
1>  rms/ios/build/device/Huntfun.ipa - {1}
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

Are there any updates/npm updates that are required on the machine with visual studio installed?

**Comment by rvfede** [comment]
Hi Guys! 
No news about this issue? 
Did you have an idea about when the fix will be available?
We can wait a couple of days, after that we should try somenthing. Do you think if we downgrade the XCode we will be able to keep going for a while?
Thanks.

**Comment by jlyshoel** [comment]
@rvfede I dont' have any news, seems that they are working on it and a fix is soon to be released. 

I can offer a workaround that worked for me. If you downgrade to XCode 8.2, you are able to build and run apps from Visual Studio. However, XCode 8.2 doesn't support iOS 10.3. If you move/copy the 10.3 "supported deviced" folder from an unpacked version of XCode 8.3, you are good to go. 

I was able to build apps using this method, but there are major problems running them from Visual Studio. Probably because xcrun doesn't really support iOS 10.3. If you click "run" in Visual Studio, your app will get deployed, but doesn't start and Visual Studio doesn't ever complete the build process. You'll have to cancel it manually and start the app manually on your device.

More on how to get Xcode 8.2 and iOS 10.3 to work together here; https://gist.github.com/steipete/d9b44d8e9f341e81414e86d7ff8fb62d

To make sure your mac is setup correctly, you can use Xcode and create a blank project to deploy. If you are able to deploy to iOS 10.3 and Xcode 8.2 it should also partly work using remotebuild.

This is the best solution I could come up with, it got me through the final stages of development and release, but it's useless if you wan't to debug and test. Hope it this may be of some help.

**Comment by rvfede** [comment]
@jlyshoel many thanks for the detailed explanation, it was really helpful, at lest I am able to build for iOS again (not yet with the iOS 10.3, I will try the work aournd later).

Now, a doubt...
If I build and send a new release to the store using XCode 8.2, will it be available for all iOS devices or 10.3 will be automatically excluded?

Thanks.

**Comment by jlyshoel** [comment]
@rvfede Glad I could be of some use. I understand your trepidation with regards to releases, and I had the same thought myself. 

I can happily report that we just released an app to App Store this morning, and a non-tech colleague of mine installed it on his iPhone with iOS 10.3 from App Store just after release. I did install it from app store on my dev device, which is also running 10.3. By that logic I'm currently working under the theory that it works for everyone, until I'm proven wrong 🥇 

With that said, I've only tried with Xcode 8.2 setup to also include support for 10.3. I'm not sure what happens if you release a build without that hack, but my assumption is that is doesn't create any problems anyway. Newer devices should in theory be able to run older code, only Xcode 8.2 that is missing a way to communicate with iOS 10.3. (My interpretation).

**Comment by AnBucyk** [comment]
@rvfede Looks likely that we'll have something released in a couple hours.

**Comment by AnBucyk** [comment]
@rvfede Quick update. We're waiting on one last set of localized strings. Worst case, we release tomorrow afternoon with a temporary gap in localization for Japanese.

**Comment by ekambarrao** [comment]
Any update about this fix!!

**Comment by rvfede** [comment]
Thanks @AnBucyk ! Will be waiting for it.

I am sure you are already aware of this, but just in case, just to let you know.
I just upgrade to VS 2017 and always when I send a remotebuild request from VS2017 it always fail with the message "The operation has timed out" BUT... if request is sent from VS2015 it works fine.

Thank you again.

**Comment by AnBucyk** [comment]
@ekambarrao We're waiting for one last set of localized strings. We're publishing today with or without them so we can get this fix out.

@rvfede There's another issue thread tracking timeouts, but I think that's VS2015 that's timing out in that thread. I can't repro 2017 timing out (my fix for this issue was tested in 2017). If you can give more details (logs, etc), feel free to add it to the existing issue.

**Comment by rvfede** [comment]
@AnBucyk, I just found that the "time out" problem I was having was related to my project and has nothing to do with the remotebuild. My fault, I am sorry.

**Comment by AnBucyk** [comment]
No worries, @rvfede 

This fix should be published. I'm closing this issue for now. :)

**Comment by DarrylLeckenby** [comment]
How is this fix being delivered? I just tried a re-install remote build via npm on the mac but the version number has not changed from my current version (2.1.3). What am I missing?

**Comment by BigFoppa** [comment]
Guess something did change after "npm install". Now I am getting this error:

"Error: Remote build error from the build server Build failed with error - 1"

**Comment by RLaptev** [comment]
Same question as @DarrylLeckenby  - how do we upgrade? 
I ran `nmp upgrade -g remotebuild`. The version is still 2.1.3 after the upgrade.

**Comment by AnBucyk** [comment]
@RLaptev @DarrylLeckenby The fix was in taco-remote-lib and taco-remote-multiplexer. These should automatically be pulled down when a new build happens. If not, try deleting ~/.taco_home/node_modules/

**Comment by RLaptev** [comment]
@AnBucyk  - thank you!
Removing ~/.taco-home/node_modules/ did the trick. The "PackageApplication" error is gone. However, I still can't build the app.... No specific error this time, just this:

`MSBUILD : cordova-build error : Remote build error from the build server http://199.204.150.52:3000/cordova - Build failed with error`

The last entries in VS2017 are not helpful:

```
...
Checking config.xml for saved plugins that haven't been added to the project
Saving ios@4.2.0 into platforms.json
No scripts found for hook "after_platform_add".
Failed to build app for buildNumber 22524:
1>Done building project "foo.jsproj" -- FAILED.
````
No specific error is provided. Do I need to enable extra logging in order to troubleshoot this?

Thank you!

**Comment by jlyshoel** [comment]
@RLaptev Got the exact same issue. I'm able to build and run directly from Xcode, but not using remotebuild+VS. I would agree that the last lines from VS isn't very helpful. Not sure if there is a verbose mode, maybe @AnBucyk could help us out here?

I did remove the taco_home directory to get remotebuild to update, which for me did removed the PackageApplication error.  I also did upgrade of xcode and xcode cli tools and reinstall of remotebuild. Didn't help. 

Just for fun, I ran "remotebuild test" on my mac, which also failed. It contains a lot more info than VS and throws an exeption from remotebuild. Thereabout is probably the root of the problem. I haven't yet figured out if it's a bug in remotebuild or something in the mac setup. Unfortenantly I won't have time to look more in detail at it today. Maybe it helps @RLaptev in your troubleshooting.

**Comment by jlyshoel** [comment]
BTW; it's the same exception mentioned here: https://github.com/Microsoft/remotebuild/issues/7

**Comment by cannvand** [comment]
@jlyshoel @RLaptev  I'm getting the exact same error, logging in VS is useless.

Prepared iOS project successfully
No scripts found for hook "after_prepare".
Checking config.xml for saved plugins that haven't been added to the project
Saving ios@4.3.1 into platforms.json
No scripts found for hook "after_platform_add".
Failed to build app for buildNumber 2469:
1>Done building project "testProject.jsproj" -- FAILED.

Any idea on how to fix this issue?

**Comment by ekambarrao** [comment]
HI, what is the current version of new release of remotebuild? I have macincloud and it has xcode 8.3.1 and ios 10.3.1 and remotebuild version 2.1.2. I am using ios@4.3.1 inside config.xml. It is giving the same error as earlier one with "packageapplication".

**Comment by AnBucyk** [comment]
@RLaptev @jlyshoel I'm not seeing the error with my setup, so I'd need a full log. When you do a build, there should be a full log saved on your Mac at ~/.taco_home/remote-builds/taco-remote/builds/<build number>/build.log. If you could send me that, I'd appreciate it. Also, since it's the same exception mentioned in #7, does the solution for that issue also work? Or is it the same exception without a solution?

@cannvand Same as above, I need a full log. You can find it at ~/.taco_home/remote-builds/taco-remote/builds/<build number>/build.log.

@ekambarrao The current version of remotebuild is 2.1.3. If you're trying to take this fix specifically, you need to clear the node_modules cached at ~/taco_home/node_modules/. This should cause remotebuild to immediately look for new versions of taco-remote-lib and taco-remote-multiplexer. Let me know if you still have issues when you try this.

**Comment by RLaptev** [comment]
@AnBucyk the log on Mac is identical to the one VS outputs. Here is the build.log
[build.log.txt](https://github.com/Microsoft/remotebuild/files/984501/build.log.txt)

Running `remotebuild test` throws this error:

```
Error: taco-remote test failed.
Build Failed: {"vcordova":"5.4.0","vcli":"1.3.5","cfg":"debug","command":"build","options":"--emulator","status":"Error","buildCommand":"build","buildPlatform":"ios","configuration":"debug","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23416","buildNumber":23416,"logLevel":0,"submissionTime":"2017-05-08T18:18:56.825Z","changeList":null,"buildSuccessful":false,"messageId":"BuildFailedWithError","messageArgs":[""],"message":"Build failed with error ","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23416/upload_23416.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23416/cordovaApp","statusTime":"2017-05-08T18:19:08.661Z","appName":"HelloCordova"}
/usr/local/lib/node_modules/remotebuild/node_modules/applicationinsights/AutoCollection/Exceptions.js:25
                    throw error;
                    ^

Error: Build Failed: {"vcordova":"5.4.0","vcli":"1.3.5","cfg":"debug","command":"build","options":"--emulator","status":"Error","buildCommand":"build","buildPlatform":"ios","configuration":"debug","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23416","buildNumber":23416,"logLevel":0,"submissionTime":"2017-05-08T18:18:56.825Z","changeList":null,"buildSuccessful":false,"messageId":"BuildFailedWithError","messageArgs":[""],"message":"Build failed with error ","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23416/upload_23416.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23416/cordovaApp","statusTime":"2017-05-08T18:19:08.661Z","appName":"HelloCordova"}
    at Request._callback (/usr/local/lib/node_modules/remotebuild/node_modules/taco-remote/lib/selftest.js:104:46)
    at Request.self.callback (/usr/local/lib/node_modules/remotebuild/node_modules/request/request.js:121:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (/usr/local/lib/node_modules/remotebuild/node_modules/request/request.js:985:14)
    at emitOne (events.js:101:20)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/remotebuild/node_modules/request/request.js:936:12)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
```

**Comment by cannvand** [comment]
@AnBucyk The log on the mac does not specify any more details. 
[build.log.txt](https://github.com/Microsoft/remotebuild/files/984516/build.log.txt)

**Comment by AnBucyk** [comment]
@RLaptev I don't believe `remotebuild test` works with Development Team ID's. It was never updated to do so, and unfortunately we're really low on manpower to implement that at the moment. So, that looks unrelated to me.

@RLaptev @cannvand Can you share the full VS side of your logs as well? I'm interested in seeing the HTTP response RemoteBuild sent to VS.

Something else that might help debug this is going to the build directory and trying to manually build the app with cordova and see what happens.

**Comment by BigFoppa** [comment]
@AnBucyk My build.log is ending the same way as those provided by @RLaptev and @cannvand. 

VS2015, Cordova 6.5.0, ios 4.4.0
macOS 10.12.4, xcode 8.3.2

**Comment by RLaptev** [comment]
@AnBucyk here is VS log:
```
-----------------------------------------------------------------------------
Sending the build to the remote build server
-----------------------------------------------------------------------------
------ Incremental Build: False
------ Submitting new build request to: http://199.204.150.52:3000/cordova/build/tasks?command=build&vcordova=6.3.1&cfg=release&options=--device
------ New Build information: {"command":"build","vcordova":"6.3.1","cfg":"release","options":"--device","status":"Uploaded","buildCommand":"build","buildPlatform":"ios","configuration":"release","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432","buildNumber":23432,"logLevel":0,"submissionTime":"2017-05-08T18:27:26.462Z","changeList":null,"buildSuccessful":false,"messageArgs":null,"message":"Uploaded build request payload.","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/upload_23432.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/cordovaApp","statusTime":"2017-05-08T18:27:43.605Z"}
------ 05/08/2017 14:27:43 Checking on build status from http://199.204.150.52:3000/cordova/build/tasks/23432 [Attempt 1]
------ New Build information: {"command":"build","vcordova":"6.3.1","cfg":"release","options":"--device","status":"Uploaded","buildCommand":"build","buildPlatform":"ios","configuration":"release","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432","buildNumber":23432,"logLevel":0,"submissionTime":"2017-05-08T18:27:26.462Z","changeList":null,"buildSuccessful":false,"messageArgs":null,"message":"Uploaded build request payload.","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/upload_23432.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/cordovaApp","statusTime":"2017-05-08T18:27:43.605Z"}
Uploaded - Uploaded build request payload.
------ 05/08/2017 14:27:48 Checking on build status from http://199.204.150.52:3000/cordova/build/tasks/23432 [Attempt 2]
------ New Build information: {"command":"build","vcordova":"6.3.1","cfg":"release","options":"--device","status":"Building","buildCommand":"build","buildPlatform":"ios","configuration":"release","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432","buildNumber":23432,"logLevel":0,"submissionTime":"2017-05-08T18:27:26.462Z","changeList":{"deletedFiles":[]},"buildSuccessful":false,"messageId":"AcquiringCordova","messageArgs":[null],"message":"Acquiring Cordova.","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/upload_23432.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/cordovaApp","statusTime":"2017-05-08T18:27:45.537Z","appName":"My Test App"}
Building - Acquiring Cordova.
------ 05/08/2017 14:27:53 Checking on build status from http://199.204.150.52:3000/cordova/build/tasks/23432 [Attempt 3]
------ New Build information: {"command":"build","vcordova":"6.3.1","cfg":"release","options":"--device","status":"Building","buildCommand":"build","buildPlatform":"ios","configuration":"release","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432","buildNumber":23432,"logLevel":0,"submissionTime":"2017-05-08T18:27:26.462Z","changeList":{"deletedFiles":[]},"buildSuccessful":false,"messageId":"UpdatingPlatform","messageArgs":["ios"],"message":"Updating platform.","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/upload_23432.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/cordovaApp","statusTime":"2017-05-08T18:27:49.762Z","appName":"My Test App"}
Building - Updating platform.
------ 05/08/2017 14:27:58 Checking on build status from http://199.204.150.52:3000/cordova/build/tasks/23432 [Attempt 4]
------ New Build information: {"command":"build","vcordova":"6.3.1","cfg":"release","options":"--device","status":"Error","buildCommand":"build","buildPlatform":"ios","configuration":"release","buildDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432","buildNumber":23432,"logLevel":0,"submissionTime":"2017-05-08T18:27:26.462Z","changeList":{"deletedFiles":[]},"buildSuccessful":false,"messageId":"BuildFailedWithError","messageArgs":[""],"message":"Build failed with error ","tgzFilePath":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/upload_23432.tgz","appDir":"/Users/mymac/.taco_home/remote-builds/taco-remote/builds/23432/cordovaApp","statusTime":"2017-05-08T18:27:56.394Z","appName":"My Test App"}
Error - Build failed with error
1>MSBUILD : cordova-build error : Remote build error from the build server http://199.204.150.52:3000/cordova - Build failed with error
------ Remote build log follows
```

**Comment by BigFoppa** [comment]
@AnBucyk here is may ending of a manual cordova build:


=== BUILD TARGET EventMap 4U OF PROJECT EventMap 4U WITH CONFIGURATION Debug ===

Check dependencies
Provisioning profile "iOS Team Provisioning Profile: io.cordova.EventMap4u" doesn't include signing certificate "iPhone Developer: Jan Forsberg (BDSH78YWJV)".
Code signing is required for product type 'Application' in SDK 'iOS 10.3'

** ARCHIVE FAILED **


The following build commands failed:
	Check dependencies
(1 failure)
Error: Error code 65 for command: xcodebuild with args: -xcconfig,/Users/ulrika/.taco_home/remote-builds/taco-remote/builds/14862/cordovaapp/platforms/ios/cordova/build-debug.xcconfig,-workspace,EventMap 4U.xcworkspace,-scheme,EventMap 4U,-configuration,Debug,-destination,generic/platform=iOS,-archivePath,EventMap 4U.xcarchive,archive,CONFIGURATION_BUILD_DIR=/Users/ulrika/.taco_home/remote-builds/taco-remote/builds/14862/cordovaapp/platforms/ios/build/device,SHARED_PRECOMPS_DIR=/Users/ulrika/.taco_home/remote-builds/taco-remote/builds/14862/cordovaapp/platforms/ios/build/sharedpch

**Comment by BigFoppa** [comment]
@AnBucyk guess we are back to the problem with code-signing. Here is what I have in build.json that has been working til now:

  "ios": {
    "debug": {
      "developmentTeam": "K3527ZMZWG",
      "codeSignIdentity": "iPhone Developer",
      "packageType": "development"
    },

    "release": {
      "developmentTeam": "K3527ZMZWG",
      "codeSignIdentity": "iPhone Developer",
      "packageType": "app-store"
    }
  }

**Comment by AnBucyk** [comment]
@BigFoppa I suspect that error's not actually related. We specifically massage your xcconfig file to include the signing and development team.

The error from the remotebuild side didn't even get to the part where it performs a cordova build ios.

Still investigating.

**Comment by AnBucyk** [comment]
@RLaptev @BigFoppa @cannvand I think I found the issue. Looks like I added an error case for this, but forgot to add a string. There's also an underlying issue with how I get the cordova-ios version (it relies on having a global install of the package). I'll have a proper fix for these issues soon, but for now, please try the following workaround:

**Try installing cordova globally on your build machine.**

Let me know if it works.

**Comment by BigFoppa** [comment]
@AnBucyk I did install cordova globally on the mac, that is why I could manually build.

I went to the Apple Developer site where you manage provisioning profiles. I added the latest certificate to each one (development and distribution) went in to xcode and in project editor turned off automatically manage signing and under manual signing imported the profiles.
Now it builds successfully but get an error on deploy to device.

**Comment by BigFoppa** [comment]
@AnBucyk I also in xcode opened Window/Devices and imported the latest profiles to the device. Didnt work. Still get the deploy error. Here is the end of the build log in VS2015:

Done building 15347
1>  ------ Downloading completed build files from http://192.168.1.10:3000/cordova/build/15347/download to K:\projekt\Mobile\fb-events\bld\iOS\Debug
1>  ------ Done downloading the build outputs to K:\projekt\Mobile\fb-events\bld\iOS\Debug
1>  ------ Extracting the build outputs to K:\projekt\Mobile\fb-events\bin\iOS\Debug
1>  ------ Done extracting the build outputs to K:\projekt\Mobile\fb-events\bin\iOS\Debug
1>  ------ Copying back to project: iOS
2>------ Deploy started: Project: fb-events, Configuration: Debug iOS ------
2>  ------ Cordova tools 6.5.0 already installed.
2>  Requesting deploy on remote iOS device for buildNumber 15347 on server http://192.168.1.10:3000/cordova...
2>Failed to deploy iOS remote for build K:\projekt\Mobile\fb-events\bld\ios\Debug\buildInfo.json to http://192.168.1.10:3000/cordova :
2>  
2>Http 404: {"stdout":"Copying '/Users/ulrika/.taco_home/remote-builds/taco-remote/builds/15347/cordovaApp/platforms/ios/build/device/EventMap 4U.ipa' to device... DONE.\nInstalling 'io.cordova.EventMap4u'\n - CreatingStagingDirectory (5%)\n\r - ExtractingPackage (15%)\n\r - InspectingPackage (20%)\n\r - TakingInstallLock (20%)\n\r - PreflightingApplication (30%)\n\r - InstallingEmbeddedProfile (30%)\n\r - VerifyingApplication (40%)\n\r - CreatingContainer (50%)\n\r - InstallingApplication (60%)\n\r - PostflightingApplication (70%)\n\r - SandboxingApplication (80%)\n\r - GeneratingApplicationMap (90%)\n","stderr":"WARNING: could not locate iTunesMetadata.plist in archive!\nWARNING: could not locate Payload/EventMap 4U.app/SC_Info/EventMap 4U.sinf in archive!\n","code":null,"message":"Error installing app to device: WARNING: could not locate iTunesMetadata.plist in archive!\nWARNING: could not locate Payload/EventMap 4U.app/SC_Info/EventMap 4U.sinf in archive!\n"}
2>Failed to deploy iOS remote for build K:\projekt\Mobile\fb-events\bld\ios\Debug\buildInfo.json to http://192.168.1.10:3000/cordova :
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Deploy: 0 succeeded, 1 failed, 0 skipped ==========

**Comment by BigFoppa** [comment]
@AnBucyk I in xcode opened the xcconfig file from the latest build and it builds, deploys and runs without any problem.

**Comment by AnBucyk** [comment]
@BigFoppa I'm not sure the current issues you're seeing have anything to do with the previous issue. Are you trying a Local Deploy?

@RLaptev @cannvand Can one of you try the workaround above and let me know if it works?

**Comment by BigFoppa** [comment]
@AnBucyk I am not sure what you mean by "Local Deploy".

**Comment by RLaptev** [comment]
@AnBucyk I run the following command on my deployment Mac:

`npm install -g cordova`

getting a new error now:

`Build failed with error Remotebuild requires your projects to use cordova-ios 4.3.0 or greater with XCode 8.3. Please update your cordova-ios version.`

Cordova was nor previously installed on this Mac as [documentation](https://docs.microsoft.com/en-us/visualstudio/cross-platform/tools-for-cordova/first-steps/ios-guide) says:

> Cordova and its dependencies are automatically acquired and installed as needed during the build process.

`cordova --version` command returns 7.0

**Comment by BigFoppa** [comment]
@RLaptev try to uninstall and install previous version
npm uninstall -g cordova
npm install -g cordova@6.5.0

But you should use later version of the ios platform as it states to fix issues with xcode 8.3.2
I am using cordova 6.5.0 and ios 4.4.0.  Cordova 7.0 is the latest release which I have not tried yet in my project.

**Comment by jlyshoel** [comment]
Got the same issue as @RLaptev, but I noticed that my project was using 6.3.1 in my project,
![project](https://cloud.githubusercontent.com/assets/25656625/25837580/ccc162fe-348e-11e7-91b7-a2beaab08fb4.png)

I then did as admin on my workstation:
```
npm install -g cordova
```

And on the Mac, based on the feedback from remotebuild
```
npm install -g ios-deploy
```

That seems to have fixed the build problems. However, my app doesn't start on the phone, so I'm sort of back to where I was a week ago running a "hacked" version og Xcode 8,2. 

@RLaptev You get the same error?

**Comment by cannvand** [comment]
@AnBucyk Worked like a charm.
Ran this command on the MAC

> sudo install -g cordova

This installed cordova 7.0.0 globaly.
Tried to build using VS and it worked.
Note on my windows workstation I have cordova 6.5.0 globaly installed.

**Comment by AnBucyk** [comment]
@cannvand Glad it worked for you. I'll have a fix soon so that you no longer need a globally installed cordova, but that should work in the meantime.

@jlyshoel You can override the versions of cordova, tools and platforms by selecting "Global" for your toolset, and then manually installing the tools you need. The idea of the 6.3.1 toolset is it's a known, good configuration that should work out-of-box. The only caveat here is that the 6.3.1 toolset doesn't support remotebuild w/ Xcode 8.3 unless you edit config.xml manually to use cordova-ios 4.3.0 or greater.

@RLaptev Great, I'm glad you're seeing that error. That's expected. The minimum cordova-ios version we support with Xcode 8.3 is now 4.3.0 or greater. As above, you should be able to manually edit your config.xml to specify cordova-ios to be 4.3.0, and your build should succeed.

@BigFoppa The error you're seeing is something I'd expect if the project is using libimobile to deploy to a device, and so far as I know, we only use that for the Local Deploy feature in Visual Studio. I could be wrong about this, though - I'm pretty new to this codebase. Regardless, the error you're getting looks unrelated. Feel free to open a new issue for this if you want us to investigate it further.

**Comment by jlyshoel** [comment]
@AnBucyk That solved the build problem, but still unable to actually run the application on the phone. It gives a HTTP 404 on /deploy

```
2>------ Deploy started: Project: Fernerbonus, Configuration: Debug iOS ------
2>Requesting deploy on remote iOS device for buildNumber 3795 on server http://10.0.11.125:3000/cordova...
2>http://10.0.11.125:3000/cordova/build/3795/deploy The remote server returned an error: (404) Not Found.
```

Any idea if this is related?

**Comment by AnBucyk** [comment]
Physical device or simulator? Sometimes it takes a moment for the simulator to boot and it 404's if you're too quick. Try again once it's booted? If it's a device, I'm not sure, and we'll need to investigate.

Regardless, it's probably not related (this fix only really effects signing/packaging, not deploying. Feel free to file a new issue.

**Comment by BigFoppa** [comment]
@AnBucyk When you say Local Deploy do you mean building for Local Device? 
No, I am building for Remote Device on my mac.

**Comment by ekambarrao** [comment]
@AnBucyk, seeking help ....this totally made me go nuts now with no clue about which version is compatible to what to get this build successful.

 I am still getting the same error. VS2015/2017, Cordova 6.5/6.3, ios 10.3, xcode 8.3.1 remotebuild 2.1.3
 ios is ~4.3.1 in config

-----------------------------------------------------------

1>MSBUILD : cordova-build error : xcrun: error: unable to find utility "PackageApplication", not a developer tool or in PATH
1>xcrun : error : unable to find utility "PackageApplication", not a developer tool or in PATH
1>MSBUILD : cordova-build error :  - 1
1>   - {1}
-----------------------------------------

if I use cordova 7.0 in vs 2015 and build it, 

I am getting another weird error

----------------------------------------------------------------------------------
 No scripts found for hook "before_platform_add".
1>  Removing "cordova-" prefix from cordova-ios
1>  Adding ios project...
1>  Error PlatformApi not loaded for platform.SyntaxError C:\Users\Ekambarrao\.cordova\lib\npm_cache\cordova-ios\4.4.0\package\bin\templates\scripts\cordova\lib\plugman\pluginHandlers.js:99
1>                          let opt = { customFramework: false, embed: false, link
1>                          ^^^
1>  Unexpected strict mode reserved word
1>MSBUILD : cordova-build error : Error: Your ios platform does not have Api.js
1>  Error Your ios platform does not have Api.js
------------------------------------------------------------------------------------------------------------

**Comment by cannvand** [comment]
@ekambarrao did you try removing the node_modules folder? this should fix the first error, the second one I cannot explain.

> @RLaptev @DarrylLeckenby The fix was in taco-remote-lib and taco-remote-multiplexer. These should automatically be pulled down when a new build happens. If not, try deleting ~/.taco-home/node_modules/

**Comment by BigFoppa** [comment]
@cannvand Did you ever see the deploy 404 error?

@AnBucyk using "cordova run --device" in the {build-number}/cordovaApp folder successfully deploys and runs the app on my iPhone 6 device.

**Comment by ekambarrao** [comment]
@cannvand, thx for help. it gave me ipa.

**Comment by cannvand** [comment]
@BigFoppa No i've never seen the deploy 404 error. Looks like the build succeeds but VS is unable to copy the .ipa.
If you browse to the folder on the mac do you see the .ipa file?

**Comment by Adavo** [comment]
Just wanted to insist that is important to:
- install latest version on windows : npm install -g cordova
- configure VS to use cordova 7.0.0 (on config.xml, in the platforms tab. If you don't see it, just change taco.json to 7.0.0)
- install latest cordova on mac : npm install -g cordova
- install on mac the latest version of ios-deploy : npm install ios-deploy
- remove the folder ~/.taco-home/node_modules/
- install the latest remotebuild on mac : npm install remotebuild (it should be 2.1.3).

And it works !

**Comment by BigFoppa** [comment]
@cannvand Yes the .ipa file is there. And it is also copied to the pc.

**Comment by RLaptev** [comment]
@AnBucyk running `npm install -g ios-deploy` on Mac solved the cordova-ios 4.3.0 version requirement error. However, I still can't fully compile:
Snippet of the error:
```
Can't find any applicable signing identities for items
Error Domain=IDECodesignResolverErrorDomain Code=10 \"No matching provisioning profiles found\" UserInfo={NSLocalizedRecoverySuggestion=No valid 'ad-hoc' profiles were found., NSLocalizedDescription=No matching provisioning profiles found,
```
My build.json contains the following:
```
  "ios": {
    "debug": {
      "developmentTeam": "XXX",
      "provisioningProfile": "64B49D46-E09B-42F4-9BDC-7F66A852C3CF"
    },
    "release": {      
      "developmentTeam": "XXX",
      "provisioningProfile": "7C831EED-C5BB-4339-9260-C04BBCC70EAD",
      "packageType": "ad-hoc"
    }
  }
```
The above error comes at what seems like almost the end of the compilation. 
I know XCode has access to this provisioning profile as it prompted me to confirm access permissions during the remote build.
If I remove provisioningProfile from build.json then compilation fails with `Could not find development team` error right from the start.
I have Remove Device as target selected in VS when doing the build. 
The installed on Mac provisioning profile is for **In House App Development**, since the app is for internal use and will not be uploaded to Apple store.

**Comment by AnBucyk** [comment]
@RLaptev I think the error you're seeing is one that I was seeing before I had a **valid** provisioning profile on the build machine. Made sure your provisioning profile is installed and your App ID applies? To be clear, this is an Apple thing, not a remotebuild thing. Personally, I had to manually download and install my provisioning profile to `~/Library/MobileDevice/Provisioning Profiles` to get it to work.

> configure VS to use cordova 7.0.0 (on config.xml, in the platforms tab. If you don't see it, just change taco.json to 7.0.0)

This is NOT the way I'd recommend doing this. Cordova 7.0.0 isn't required (or officially supported yet). If you want to use it, I highly suggest switching your toolset to Global and installing your dependencies manually (Including Cordova).

@BigFoppa Please open a new issue for your 404 error - I'm not going to try to debug it in the comment thread for an unrelated issue anymore.

Everyone else - Please try the following things if you're still having issues. Anything unrelated specifically to this issue should have its own issue thread - feel free to create one.

- Clear `~/.taco_home/node_modules`
- Install cordova (any version) globally on your build machine (Temporary workaround, fix coming soon)
- If you're using Xcode 8.3, ensure the cordova-ios platform version is 4.3.0 or newer.

**Comment by gubyy** [comment]
@Adavo I did these steps but now getting this error on VS15:

Error: Your ios platform does not have Api.js

**Comment by RLaptev** [comment]
@AnBucyk the profiles were in the right places. I had to add packageType to build.json to finally compile the app. All good now!

  ```
"ios": {
    "debug": {
      "developmentTeam": "XXX",
      "provisioningProfile": "64B49D46-E09B-42F4-9BDC-7F66A852C3CF"
    },
    "release": {
      "codeSignIdentity": "iPhone Distribution",
      "developmentTeam": "XXX",
      "provisioningProfile": "7C831EED-C5BB-4339-9260-C04BBCC70EAD",
      "packageType": "development"
    }
  }
```

**Comment by BigFoppa** [comment]
@AnBucyk Thank you for your efforts! At least I can build now and run with xcode. 
I totally understand that you want to separate my deploy issue from your work with the build. I will later file an issue for this, just a bit busy right now (aren't we all :-)...

@RLaptev I believe that "packageType" for release should be "app-store" and for "debug" it should be "development".

**Comment by BigFoppa** [comment]
And @Adavo installing Cordova 7 on my machine with VS2015 was a mistake. All hell broke loose, especially trying to build for android. Guess many plugins are not ready for this upgrade. Took me a while to clean up and be back to square 1.

**Comment by RLaptev** [comment]
@BigFoppa - in my case for in-house app packageType needs to be enterprise for release. thanks.

**Comment by ekambarrao** [comment]
Hi @AnBucyk , I have Visual Studio 2017 and used Cordova 7.0 and ios 8.3.1 and Mac 10.X. I am able to successfully build the solution and getting the ipa and plist. Now I am held up at "Missing info.plist" and I have gone through all the link available and does all mentioned in several posts found it be not resolving the issue. Any suggestion? Note: I know that this not remotebuild issue to add here in this 'thread'

**Comment by BigFoppa** [comment]
@ekambarrao found this on stackoverflow: http://stackoverflow.com/questions/39980700/usage-description-issue-in-ionic-and-ios-10-builds

Hope it helps you...

**Comment by ekambarrao** [comment]
I have used all options. like reinstalled the camera plugin and then included the config.xml tags etc. they are reflecting inside *-info.plist but when submitting to App Store, still getting the same error.

**Comment by ekambarrao** [comment]
<plugin src="https://github.com/apache/cordova-plugin-media-capture.git" name="cordova-plugin-media-capture" spec="~1.4.0">
<variable value="capture image to add to message" name="CAMERA_USAGE_DESCRIPTION"/>
<variable value="get image to add to message" name="MICROPHONE_USAGE_DESCRIPTION"/>
<variable value="get image to add to message" name="PHOTOLIBRARY_USAGE_DESCRIPTION"/>
</plugin>
<plugin name="cordova-plugin-queries-schemes" spec="https://github.com/rjhilgefort/cordova-plugin-queries-schemes.git"/>
<plugin src="https://github.com/rjhilgefort/cordova-plugin-queries-schemes.git" name="cordova-plugin-queries-schemes" spec="~0.1.1"/>
<engine name="ios" spec="~4.4.0"/>
<plugin src="https://github.com/dpa99c/cordova-custom-config.git" name="cordova-custom-config" spec="~3.2.0"/>
<plugin src="https://github.com/apache/cordova-plugin-camera.git" name="cordova-plugin-camera" spec="~2.3.0">
<variable value="capture image to add to message" name="CAMERA_USAGE_DESCRIPTION"/>
<variable value="get image to add to message" name="PHOTOLIBRARY_USAGE_DESCRIPTION"/>
</plugin>

<preference name="CAMERA_USAGE_DESCRIPTION" default=" "/>
<config-file parent="NSCameraUsageDescription" target="*-Info.plist" platform="ios">
<string>$CAMERA_USAGE_DESCRIPTION</string>
</config-file>
<preference name="MICROPHONE_USAGE_DESCRIPTION" default=" "/>
<config-file parent="NSMicrophoneUsageDescription" target="*-Info.plist" platform="ios">
<string>$MICROPHONE_USAGE_DESCRIPTION</string>
</config-file>
<preference name="PHOTOLIBRARY_USAGE_DESCRIPTION" default=" "/>
<config-file parent="NSPhotoLibraryUsageDescription" target="*-Info.plist" platform="ios">
<string>$PHOTOLIBRARY_USAGE_DESCRIPTION</string>
</config-file>

**Comment by stevecasey17** [comment]
Had this issue and to get past the remote build error had to open the remote build project in xcode on the remote build mac. 
The signing certs were not correct. 
Selected the correct signing certs in xcode. 
Build and deployed via xcode without issue.
The remote build works now that the signing cert is correct. 
Compared files on remote vs local build. No changes.
At least it is building , debug is another story.
